x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+                name(),
+                lowerTerm == null ? null : lower,
+                upperTerm == null ? null : upper,
+                includeLower,
+                includeUpper
+            );
+
+            return new BooleanQuery.Builder().add(new BooleanClause(pointPrefixQuery, Occur.MUST))
+                .add(new BooleanClause(termQuery, Occur.MUST))
+                .build();
+        }
+    }
+
+    private String nullValue;
+    private BooleanFieldMapper prereleaseSubField;
+    private NumberFieldMapper majorVersionSubField;
+    private NumberFieldMapper minorVersionSubField;
+    private NumberFieldMapper patchVersionSubField;
+
+    private VersionStringFieldMapper(
+        String simpleName,
+        FieldType fieldType,
+        MappedFieldType mappedFieldType,
+        String nullValue,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        BooleanFieldMapper preReleaseMapper,
+        NumberFieldMapper majorVersionMapper,
+        NumberFieldMapper minorVersionMapper,
+        NumberFieldMapper patchVersionMapper
+    ) {
+        super(simpleName, fieldType, mappedFieldType, multiFields, copyTo);
+        this.nullValue = nullValue;
+        this.prereleaseSubField = preReleaseMapper;
+        this.majorVersionSubField = majorVersionMapper;
+        this.minorVersionSubField = minorVersionMapper;
+        this.patchVersionSubField = patchVersionMapper;
+    }
+
+    @Override
+    public VersionStringFieldType fieldType() {
+        return (VersionStringFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected VersionStringFieldMapper clone() {
+        return (VersionStringFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String versionString;
+        if (context.externalValueSet()) {
+            versionString = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                versionString = nullValue;
+            } else {
+                versionString = parser.textOrNull();
+            }
+        }
+
+        if (versionString == null) {
+            return;
+        }
+
+        EncodedVersion encoding = encodeVersion(versionString);
+        BytesRef encodedVersion = encoding.bytesRef;
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            Field field = new Field(fieldType().name(), encodedVersion, fieldType);
+            context.doc().add(field);
+            // encode the first 16 bytes as points for efficient range query
+            byte[] first16bytes = Arrays.copyOfRange(encodedVersion.bytes, encodedVersion.offset, 16);
+            context.doc().add(new BinaryPoint(fieldType().name(), first16bytes));
```

Comment:
Should we index `BinaryPoint` only when `indexOptions() != IndexOptions.NONE `(because currently this field will be indexed even when `IndexOptions.NONE` and stored=true) ?